Simplicity and validation

zerosonesfun opened the pull request:
These changes make things simple, ensure the feeds validate, and keep the feed post content text only versus trying to include HTML. This last part is admittedly a personal preference. I do not think we need the feeds to include the full HTML for posts. Plain text descriptions/summaries should suffice and help to ensure validation and feed reader compatibility.

The main reason I started making these changes is because the ID portion of the atom feed wasn’t validating. It was only pulling in the post ID which happens to be “1” every time. This pull request makes the ID the full permalink which validates.

zerosonesfun committed:
Update atom.blade.php

zerosonesfun committed:
Update rss.blade.php

dexif commented:
+1

AmauryCarrade commented:
Hi,

Thank you for your PR! I'm not home right now so I won't be able to review and test it until next week, but it's very appreciated :)

zerosonesfun committed:
Update README.md

AmauryCarrade commented on README.md:
```diff
@@ -20,4 +20,4 @@ This extension adds the following feeds to Flarum:
 You can replace `atom` by `rss` in the URLs above to get RSS feeds instead. The tags-related feeds are only available if the tags extension is installed and enabled.
 
-Feeds are linked in the pages for autodiscovery and listing in the browser (at least in Firefox, as Chrome, Vivaldi, and a lot of others doesn't support syndication feeds out of the box). This said, they are not dynamically updated as the page change (except when fully reloaded), becuse of this [Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=380639). If a workaround is found, they will ...
+The difference between this fork and the original is that this fork has tweaked a couple things to ensure the feeds validate.
```
I suppose you added this for your fork, but can you revert in order to merge your changes?

zerosonesfun replied:
I reverted the readme file back to your original.

AmauryCarrade commented on views/atom.blade.php:
```diff
@@ -13,10 +13,9 @@
-        <id>{{ $entry['id'] or $entry['permalink'] }}</id>
+        <id>{{ $entry['permalink'] }}</id>
         <updated>{{ $entry['pubdate']->format(DateTime::ATOM) }}</updated>
         <summary><![CDATA[{!! $entry['description'] !!}]]></summary>
```
As posts can include HTML, we should add an attribute here for the feed to validate.
```suggestion
        <summary type="html"><![CDATA[{!! $entry['description'] !!}]]></summary>
```

zerosonesfun replied:
The HTML is not included if you stick with `$entry['description']`. It seems like the HTML is only included if you use `$entry['content']`. Which is currently in your code. I removed it in my fork because I do not want the HTML. I only want a plain text feed. You may test my feed which uses my forked version of this code by plugging the following address into a feed validator: https://ennuily.com

Some posts at this forum have HTML. But, you will see no HTML in my feed. 

I guess this is a personal preference. Some may want HTML in their feeds. I do not which is why I went with the description only versus the content. This is a decision which is up to you. Do you want HTML in the feed summary? If you do, “description” would need to be changed to “content” and then you could add the type=html. But, again I prefer text only excerpts/summaries and so I would stick with using my forked version if you were to decide to go with “content” versus “description.”

AmauryCarrade replied:
I guess too that's a personal preference. To be honest, I'm on the « full+html feed » side, to be able to consume the whole content from an aggregator without having to open the forum. But I understand that's not the prefered behaviour of everyone. I could for the moment use your solution (so I don't add even more time to this upgrade) and later add an option between text and HTML; and between full content and excerpts.

zerosonesfun replied:
Yes, the ideal solution would be admin settings. This way people can switch from full content with HTML to just text only excerpts. You could have 4 blade.php files. Two RSS and two Atom. The admin setting would switch between these depending on the HTML/non-HTML preference. 

It’s your extension so if your preference is full content / full HTML I say go for it. 🙂 You could accept and merge this pull request which fixes a couple things, and then change the “description” code to “content” and add that HTML attribute where needed.

Another option - no admin settings needed... copy the RSS and Atom blade files and give them a different name. By default the extension could use the full HTML versions. But, in the install instructions you tell people that they can simply rename the non-HTML versions of the files to the appropriate names, and either delete or rename the full HTML files. It’s manual, but gives people a way to somewhat easily switch between the two options. 

Actually, I’ll go ahead and do this so you see what I mean. I can undo it if you don’t like the idea.

AmauryCarrade commented on views/rss.blade.php:
```diff
@@ -13,9 +13,7 @@
-            <content:encoded><![CDATA[{!! $entry['content'] or $entry['description'] !!}]]></content:encoded>
-            <link>{{ $entry['permalink'] }}</link>
-            <guid isPermaLink="{{ !isset($entry['id']) }}">{{ $entry['id'] or $entry['permalink'] }}</guid>
+            <guid>{{ $entry['permalink'] }}</guid>
```
We may want to add `isPermalin="true"` here.

```suggestion
            <guid isPermaLink="true">{{ $entry['permalink'] }}</guid>
```

zerosonesfun replied:
If the guid is a full URL, which this is, you don’t need the isPermaLink. See here: https://validator.w3.org/feed/docs/error/InvalidPermalink.html

AmauryCarrade replied:
You're right, I misread the spec.

AmauryCarrade reviewed (changes requested):
H, and again sorry for the delay. I added a few comments to be addressed before the merge, but thanks a lot for the PR :) .

zerosonesfun committed:
Update README.md

zerosonesfun committed:
Update atom.blade.php

zerosonesfun committed:
Update atom.blade.php

zerosonesfun commented:
I noticed another small issue when validating the atom feed. It did not like the site URL not ending with a “/“. Therefore, I added a slash in the code. 

There is just one more warning at times but it is not a big deal. Sometimes, depending on the post (a post with no text, just an image, for example), the summary may end up being blank. The feed validator still marks the feed as valid, but gives a warning that there shouldn’t be blank summaries. In the future, some type of if/else coding could solve this... if summary is blank, add the text, “No summary found.” This way, there is never a blank summary.

Here are screenshots of my feeds validating. As I state above in my replies, the couple changes you suggest are not necessary; the feeds are valid (you will see a warning about the utf-8 encoding but that’s just something with my server. Others may not get that warning...

RSS valid:

<img src="https://i.ibb.co/mrp07yY/0600-BE66-232-F-4106-9-F66-E108-C1-B75978.jpg" alt="0600-BE66-232-F-4106-9-F66-E108-C1-B75978" border="0">

ATOM valid:

<img src="https://i.ibb.co/Qn2fkTW/609132-F1-5-D56-4-F67-986-B-833-A0-FF00-D31.jpg" alt="609132-F1-5-D56-4-F67-986-B-833-A0-FF00-D31" border="0">

zerosonesfun committed:
Update views/atom.blade.php

Co-Authored-By: Amaury Carrade <amaury@carrade.eu>

zerosonesfun committed:
Update atom.blade.php

Added html attribute and changed summary to full HTML content.

zerosonesfun committed:
Update rss.blade.php

Changed description to full HTML content.

zerosonesfun committed:
Create atom.blade.nonHTML.php

zerosonesfun committed:
Create rss.blade.nonHTML.php

zerosonesfun committed:
Update atom.blade.php

zerosonesfun commented:
@AmauryCarrade,

Please review my latest changes. Here is what I have done:

- Changed atom.blade.php so that it will now include the full HTML content of the post; and it includes the type=html attribute.
- Changed rss.blade.php so that it will also include the full HTML content of the post.
   - By default, the extension will use these full HTML versions.
- Added a new atom.blade.nonHTML.php file and a rss.blade.nonHTML.php file which are the same feeds except they will only show plain text excerpts of the posts.
   - This means, if users want, they can remove or rename the original blade.php files, and remove the "nonHTML" part in these filenames and then the extension will use the non HTML / excerpt versions instead.

AmauryCarrade commented:
> This means, if users want, they can remove or rename the original blade.php files, and remove the "nonHTML" part in these filenames and then the extension will use the non HTML / excerpt versions instead.

One should never change anything in the `vendor` directory, so I'm not convinced. This being said, I'll merge these changes and add a settings interface to configure:

- full text vs extract;
- plain vs HTML;
- how many entries per feed.

Thanks for your contributions!
